modules/imgcodecs/src/grfmt_spng.cpp

```diff
@@ -0,0 +1,486 @@
+                        {
+                            ret = spng_get_row_info(png_ptr, &row_info);
+                            if(ret) break;
+
+                            ret = spng_decode_row(png_ptr, image + row_info.row_num * image_width, image_width);
+                            if(fmt == SPNG_FMT_RGB8) {
+                                icvCvt_BGR2Gray_8u_C3C1R(
+                                        reinterpret_cast<const uchar *>(image + row_info.row_num * image_width),
+                                        0,
+                                        img.data + row_info.row_num * (image_width/3) ,
+                                        0, Size(static_cast<int>(image_width)/3,1), 2);
+                            } else if(fmt == SPNG_FMT_RGBA8) {
+                                icvCvt_BGRA2Gray_8u_C4C1R(
+                                        reinterpret_cast<const uchar *>(image + row_info.row_num * image_width),
+                                        0,
+                                        img.data + row_info.row_num * (image_width/4) ,
+                                        0, Size(static_cast<int>(image_width)/4,1), 2);
+                            } else {
+                                icvCvt_BGRA2Gray_16u_CnC1R(
+                                        reinterpret_cast<const ushort *>(image + row_info.row_num * image_width), 0,
+                                        reinterpret_cast<ushort *>(img.data + row_info.row_num * (image_width / 4)), 0 , Size(static_cast<int>(image_width) / 4, 1), 3, 2);
+                            }
+                        }
+                        while(!ret);
+                    } else {
+                        do
+                        {
+                            ret = spng_get_row_info(png_ptr, &row_info);
+                            if(ret) break;
+
+                            ret = spng_decode_row(png_ptr, img.data + row_info.row_num * image_width, image_width);
+                        }
+                        while(!ret);
+                    }
+
+                    if(ret == SPNG_EOI) {
+
+                        if(color) {
+                            // RGB -> BGR
+                            unsigned char temp;
+                            for(int i = 0;  i < img.cols * img.rows * img.channels();  i += img.channels() )
+                                std::swap(img.data[i], img.data[i+2]);
+                        }
+
+#ifdef PNG_eXIf_SUPPORTED
+                        struct spng_exif exif_s;
+                        ret = spng_get_exif(png_ptr, &exif_s);
+                        if(!ret) {
+                            if (exif_s.data && exif_s.length > 0) {
+                                m_exif.parseExif((unsigned char *) exif_s.data, exif_s.length);
+                            }
+                        }
+#endif
+                        result = true;
+                    }
+                }
+
+            }
+        }
+
+        close();
```

Comment:
the API is unstable. You call close() after data is read. User may want, for some reason, call readData several times.

Reply:
I changed it to
```cpp
        if(!result)
            close();
```

Should I create a quick patch for the current PNG decoder too?